File: packages/redisutil/src/redisutil/key.py
from __future__ import annotations

from dataclasses import dataclass
import re
from urllib.parse import quote


MAX_KEY_LENGTH = 512
_STRUCTURAL_SEGMENT = re.compile(r"^[A-Za-z0-9][A-Za-z0-9._-]*$")


@dataclass(frozen=True, slots=True)
class KeyBuilder:
    """Build namespaced Redis keys without owning a Redis client.

    Example:
        >>> builder = KeyBuilder("local", "product-service", 1)
        >>> builder.build("catalog", "product/42")
        'local:product-service:v1:catalog:product%2F42'
        >>> builder.build_with_hash_tag(
        ...     "category:7", "catalog", "product/42"
        ... )
        'local:product-service:v1:{category%3A7}:catalog:product%2F42'
    """

    environment: str
    service: str
    schema_version: int

    def __post_init__(self) -> None:
        """Validate the immutable namespace used by every generated key."""
        _validate_structural_segment("environment", self.environment)
        _validate_structural_segment("service", self.service)
        if self.schema_version < 1:
            raise ValueError("schema version must be positive")

    def build(
        self,
        *identifiers: str,
    ) -> str:
        """Build a key using the shared environment/service/version contract.

        Args:
            identifiers: One or more dynamic key segments.

        Returns:
            A percent-encoded Redis key.

        Raises:
            ValueError: If a segment or the complete key is invalid.
        """
        if not identifiers:
            raise ValueError("at least one identifier is required")
        return self._build(None, identifiers)

    def build_with_hash_tag(
        self,
        hash_tag: str,
        *identifiers: str,
    ) -> str:
        """Build a key with an explicit Redis Cluster hash tag.

        Args:
            hash_tag: Dynamic segment used for Redis Cluster slot selection.
            identifiers: Additional dynamic key segments.

        Returns:
            A percent-encoded Redis key containing one explicit hash tag.

        Raises:
            ValueError: If a segment or the complete key is invalid.
        """
        encoded_hash_tag = _encode_identifier("hash_tag", hash_tag)
        return self._build(
            f"{{{encoded_hash_tag}}}",
            identifiers,
        )

    def _build(
        self,
        hash_tag: str | None,
        identifiers: tuple[str, ...],
    ) -> str:
        segments = [
            self.environment,
            self.service,
            f"v{self.schema_version}",
        ]
        if hash_tag is not None:
            segments.append(hash_tag)
        segments.extend(
            _encode_identifier(f"identifier_{index}", identifier)
            for index, identifier in enumerate(identifiers)
        )
        key = ":".join(segments)
        key_length = len(key.encode("utf-8"))
        if key_length > MAX_KEY_LENGTH:
            raise ValueError(
                f"key length {key_length} exceeds maximum {MAX_KEY_LENGTH}"
            )
        return key


def _validate_structural_segment(name: str, value: str) -> None:
    if not _STRUCTURAL_SEGMENT.fullmatch(value):
        raise ValueError(
            f"{name} must contain only letters, numbers, dot, underscore, or hyphen"
        )


def _encode_identifier(name: str, value: str) -> str:
    if not value:
        raise ValueError(f"{name} is required")
    return quote(value, safe="-._~")

File: packages/go-platform/redisutil/key.go
package redisutil

import (
	"net/url"
	"regexp"
	"strconv"
	"strings"

	"github.com/samber/oops"
)

const MaxKeyLength = 512

var (
	keyErr            = oops.In("platform_redis_key").Code("redis.key_invalid")
	structuralSegment = regexp.MustCompile(`^[A-Za-z0-9][A-Za-z0-9._-]*$`)
)

// KeyBuilder creates namespaced Redis keys without owning a Redis client.
//
// Example:
//
//	builder, err := NewKeyBuilder("local", "product-service", 1)
//	key, err := builder.Build("catalog", "product/42")
//	// key == "local:product-service:v1:catalog:product%2F42"
//
// Use BuildWithHashTag only when related Redis Cluster keys must share a slot:
//
//	key, err = builder.BuildWithHashTag("category:7", "catalog", "product/42")
//	// key == "local:product-service:v1:{category%3A7}:catalog:product%2F42"
type KeyBuilder struct {
	prefix string
}

func NewKeyBuilder(environment, service string, schemaVersion int) (KeyBuilder, error) {
	if err := validateStructuralSegment("environment", environment); err != nil {
		return KeyBuilder{}, err
	}
	if err := validateStructuralSegment("service", service); err != nil {
		return KeyBuilder{}, err
	}
	if schemaVersion < 1 {
		return KeyBuilder{}, keyErr.With("segment", "schema_version").New("schema version must be positive")
	}
	return KeyBuilder{
		prefix: environment + ":" + service + ":v" + strconv.Itoa(schemaVersion),
	}, nil
}

func (b KeyBuilder) Build(identifiers ...string) (string, error) {
	if len(identifiers) == 0 {
		return "", keyErr.With("segment", "identifier").New("at least one identifier is required")
	}
	return b.build("", identifiers)
}

func (b KeyBuilder) BuildWithHashTag(
	hashTag string,
	identifiers ...string,
) (string, error) {
	encodedHashTag, err := encodeIdentifier("hash_tag", hashTag)
	if err != nil {
		return "", err
	}
	return b.build("{"+encodedHashTag+"}", identifiers)
}

func (b KeyBuilder) build(
	hashTag string,
	identifiers []string,
) (string, error) {
	if b.prefix == "" {
		return "", keyErr.With("segment", "prefix").New("key builder is not initialized")
	}
	segments := []string{b.prefix}
	if hashTag != "" {
		segments = append(segments, hashTag)
	}
	for index, identifier := range identifiers {
		encoded, err := encodeIdentifier("identifier_"+strconv.Itoa(index), identifier)
		if err != nil {
			return "", err
		}
		segments = append(segments, encoded)
	}
	key := strings.Join(segments, ":")
	if len(key) > MaxKeyLength {
		return "", keyErr.With("length", len(key), "max_length", MaxKeyLength).New("key is too long")
	}
	return key, nil
}

func validateStructuralSegment(name, value string) error {
	if !structuralSegment.MatchString(value) {
		return keyErr.With("segment", name).New("segment must contain only letters, numbers, dot, underscore, or hyphen")
	}
	return nil
}

func encodeIdentifier(name, value string) (string, error) {
	if value == "" {
		return "", keyErr.With("segment", name).New("identifier is required")
	}
	return strings.ReplaceAll(url.QueryEscape(value), "+", "%20"), nil
}
